### PR TITLE
[client] Added a smoke test for changing publishing option

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation
Added a bunch of helper functions in client to generate transaction to change VMPublishingOption.

Added a smoke test for changing publishing option. Currently VMValidator didn't restart on epoch change, thus we need to manually restart the node to have a new VMValidator instance with the new modified config.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Related PRs

#2932

### CODE REVIEW

File: client/cli/src/client_proxy.rs [hunk 1]
```diff
@@ -329,6 +331,48 @@ impl ClientProxy {
         }
     }
 
+    /// Allow executing arbitrary script in the network.
+    pub fn allow_custom_script(
```
Comment (sblackshear) [inline client/cli/src/client_proxy.rs]: Nit: `enable_custom_scripts` might be slightly clearer; `allow_custom_script` sounds like you might be allowing a single custom script.

File: client/cli/src/client_proxy.rs [hunk 2]
```diff
@@ -329,6 +331,48 @@ impl ClientProxy {
...
+                    VMPublishingOption::CustomScripts,
+                ),
+                is_blocking,
+            ),
+            None => unimplemented!(),
+        }
+    }
+
+    /// Only allow executing predefined script in the network.
+    pub fn disallow_custom_script(
```
Comment (sblackshear) [inline client/cli/src/client_proxy.rs]: Similar: `disable_custom_scripts`?

Also: should this take the whitelist as an argument and/or clarify the comment to explain what "predefined" means?

File: client/cli/src/client_proxy.rs [hunk 3]
```diff
@@ -434,6 +478,7 @@ impl ClientProxy {
                 }
                 Err(e) => {
                     println!("Response with error: {:?}", e);
+                    break;
```
Comment (sblackshear) [inline client/cli/src/client_proxy.rs]: Is this change related?

File: testsuite/tests/libratest/smoke_test.rs [hunk 4]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
     );
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: Perhaps a question for later, but: should we add more test files instead of letting smoke_test grow to 1K+ lines?

File: testsuite/tests/libratest/smoke_test.rs [hunk 5]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
...
 
+#[test]
+fn test_e2e_modify_publishing_option() {
+    let (mut env, mut client_proxy) = setup_swarm_and_client_proxy(1, 0);
+    client_proxy.create_next_account(false).unwrap();
+
+    client_proxy
+        .mint_coins(&["mintb", "0", "10"], true)
+        .unwrap();
+    assert_eq!(
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: Do we need to test this?

File: testsuite/tests/libratest/smoke_test.rs [hunk 6]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
...
+        Decimal::from_f64(10.0),
+        Decimal::from_str(&client_proxy.get_balance(&["b", "0"]).unwrap()).ok()
+    );
+    let script_path = workspace_builder::workspace_root()
+        .join("testsuite/tests/libratest/dev_modules/test_script.mvir");
+    let unwrapped_script_path = script_path.to_str().unwrap();
+    let script_params = &["execute", "0", unwrapped_script_path, "script"];
+    let script_compiled_path = client_proxy.compile_program(script_params).unwrap();
+
+    // Initially publishing option was set to CustomScript, this transaction should be executed.
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: Where does this setup happen? Is it the default in `setup_swarm_and_client_proxy`?

File: testsuite/tests/libratest/smoke_test.rs [hunk 7]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
...
+            .get_sequence_number(&["sequence", "0", "true"])
+            .unwrap(),
+        1
+    );
+
+    client_proxy
+        .disallow_custom_script(&["disallow_custom_script"], true)
+        .unwrap();
+
+    // TODO: Currently VMValidator didn't restart after reconfiguration. We will manually restart
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: Is there a task for fixing the restart behavior that we can link to?

File: testsuite/tests/libratest/smoke_test.rs [hunk 8]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
...
+    let peer_to_restart = 0;
+    // restart node
+    env.validator_swarm.kill_node(peer_to_restart);
+    assert!(env
+        .validator_swarm
+        .add_node(peer_to_restart, RoleType::Validator, false)
+        .is_ok());
+
+    // mint another 10 coins after remove node 0
+    client_proxy
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: Can we avoid the extra code here by minting more earlier?

File: testsuite/tests/libratest/smoke_test.rs [hunk 9]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
...
+    //       the node so that VMValidator is using the new config.
+    let peer_to_restart = 0;
+    // restart node
+    env.validator_swarm.kill_node(peer_to_restart);
+    assert!(env
+        .validator_swarm
+        .add_node(peer_to_restart, RoleType::Validator, false)
+        .is_ok());
+
+    // mint another 10 coins after remove node 0
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: What does this comment mean?

File: testsuite/tests/libratest/smoke_test.rs [hunk 10]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
...
+            .unwrap(),
+        1
+    );
+
+    client_proxy
+        .disallow_custom_script(&["disallow_custom_script"], true)
+        .unwrap();
+
+    // TODO: Currently VMValidator didn't restart after reconfiguration. We will manually restart
+    //       the node so that VMValidator is using the new config.
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: Even though `VMValidator` will let the bad script into mempool, shouldn't trying to actually execute result in an abort? I guess I understand why we would need to do the restart if we were enabling custom scripts, but it doesn't seem like we should need it for disabling.

File: testsuite/tests/libratest/smoke_test.rs [hunk 11]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
...
+        Decimal::from_f64(20.0),
+        Decimal::from_str(&client_proxy.get_balance(&["b", "0"]).unwrap()).ok()
+    );
+
+    // Now that publishing option was changed to locked, this transaction will be rejected.
+    client_proxy
+        .execute_script(&["execute", "0", &script_compiled_path[..], "10", "0x0"])
+        .unwrap_err();
+
+    assert_eq!(
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: Comment here to say what we're checking (sequence # is still bumped by invalid script)?

File: testsuite/tests/libratest/smoke_test.rs [hunk 12]
```diff
@@ -930,6 +930,72 @@ fn test_e2e_reconfiguration() {
...
+        .unwrap();
+    assert_eq!(
+        Decimal::from_f64(20.0),
+        Decimal::from_str(&client_proxy.get_balance(&["b", "0"]).unwrap()).ok()
+    );
+
+    // Now that publishing option was changed to locked, this transaction will be rejected.
+    client_proxy
+        .execute_script(&["execute", "0", &script_compiled_path[..], "10", "0x0"])
+        .unwrap_err();
```
Comment (sblackshear) [inline testsuite/tests/libratest/smoke_test.rs]: Can we check that this error is the expected one `UNKNOWN_SCRIPT`?  It's easy to imagine this accidentally failing for the wrong reason (argument type mismatch, out of gas, etc.)

File: testsuite/tests/libratest/smoke_test.rs (hunk 10 above)
Comment (runtian-zhou) [reply testsuite/tests/libratest/smoke_test.rs]: The problem is as following: it's a bit hard to prove that this transaction is NOT committed. So if we just let the transaction through, the client will actually panic with a timeout as this transaction is never committed.

File: testsuite/tests/libratest/smoke_test.rs (hunk 5 above)
Comment (runtian-zhou) [reply testsuite/tests/libratest/smoke_test.rs]: Minting is to prove regular transaction can be processed successfully. Do you think we need it?

Comment (sblackshear) [reply testsuite/tests/libratest/smoke_test.rs]: I don't think so--this should be covered by a simpler test that ensures that you can mint from the default configuration.

File: testsuite/tests/libratest/smoke_test.rs (hunk 10 above)
Comment (sblackshear) [reply testsuite/tests/libratest/smoke_test.rs]: I'm confused. Shouldn't this transaction be committed regardless of what the script setting is? In my understanding, the only difference is that we'll see an abort if custom scripts is disabled (`UNKNOWN_SCRIPT`) and a successfully executed transaction if custom scripts is enabled.

Comment (runtian-zhou) [reply testsuite/tests/libratest/smoke_test.rs]: There's some slight difference here:
In the case when VMValidator didn't restart, this transaction will pass the VMValidator. Thus the request to submit a transaction will return with a success status. This transaction will then flow to Executor and will be dropped there. The client here will thus have no idea if this transaction is indeed dropped as this UNKNOWN_SCRIPT error will not be persisted and cannot be queried.

In the case when VMValidator restart, however, the VMValidator will reject this transaction, thus the request to submit a transaction will return with a UNKNOWN_SCRIPT error.

File: client/cli/src/dev_commands.rs [hunk 13]
```diff
@@ -115,6 +115,60 @@ impl Command for DevCommandExecute {
     }
 }
 
+pub struct DevCommandAllowCustomScript {}
```
Comment (zekun000) [inline client/cli/src/dev_commands.rs]: nit: the name probably should match the function, s/allow/enable

File: client/cli/src/dev_commands.rs [hunk 14]
```diff
@@ -115,6 +115,60 @@ impl Command for DevCommandExecute {
...
+            return;
+        }
+        match client.enable_custom_script(params, true) {
+            Ok(_) => println!("Successfully finished execution"),
+            Err(e) => println!("{}", e),
+        }
+    }
+}
+
+pub struct DevCommandDisallowCustomScript {}
```
Comment (zekun000) [inline client/cli/src/dev_commands.rs]: ditto

File: testsuite/tests/libratest/smoke_test.rs (hunk 10 above)
Comment (sblackshear) [reply testsuite/tests/libratest/smoke_test.rs]: Ah, thanks for the clarification! I think my confusion results from the following:

Why does the executor discard an `UNKNOWN_SCRIPT` transaction?

Comment (sblackshear) [reply testsuite/tests/libratest/smoke_test.rs]: Ah, I see that my expectations are wrong and `UNKNOWN_SCRIPT` is actually a `Discard`. Will look into fixing this separately. Your explanation makes sense, thanks!
